frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/toolbar/DiscoverAction.java

```diff
@@ -81,40 +84,52 @@ public void actionPerformed(AnActionEvent anActionEvent) {
-    GuiUtils.invokeLaterIfNeeded(() -> {
-      var file = GitVfsUtils.getMacheteFile(gitRepository);
-      if (file.isDefined()) {
-        OpenFileAction.openFile(file.get(), project);
-      } else {
-        VcsNotifier.getInstance(project)
-            .notifyError(
-                /* title */ getString("action.GitMachete.OpenMacheteFileAction.notification.title.machete-file-not-found"),
-                /* message */ format(
-                    getString("action.GitMachete.OpenMacheteFileAction.notification.message.machete-file-not-found"),
-                    gitRepository.getRoot().getPath()));
-      }
-    }, NON_MODAL);
+    var file = GitVfsUtils.getMacheteFile(gitRepository);
+    if (file.isDefined()) {
+      OpenFileAction.openFile(file.get(), project);
+    } else {
+      VcsNotifier.getInstance(project)
+          .notifyError(
+              /* title */ getString("action.GitMachete.OpenMacheteFileAction.notification.title.machete-file-not-found"),
+              /* message */ format(
+                  getString("action.GitMachete.OpenMacheteFileAction.notification.message.machete-file-not-found"),
+                  gitRepository.getRoot().getPath()));
+    }
   }
 
   private void saveDiscoveredLayout(IGitMacheteRepositorySnapshot repositorySnapshot, Path macheteFilePath, Project project,
-      BaseEnhancedGraphTable baseEnhancedGraphTable, IBranchLayoutWriter branchLayoutWriter) {
+      BaseEnhancedGraphTable baseEnhancedGraphTable, IBranchLayoutWriter branchLayoutWriter, @UI Runnable postWriteRunnable) {
     var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
     if (branchLayout == null) {
       VcsNotifier.getInstance(project).notifyError(
           /* title */ getString("action.GitMachete.DiscoverAction.notification.title.cannot-discover-layout-error"),
           /* message */ "");
       return;
     }
+
     new Task.Backgroundable(project, getString("action.GitMachete.DiscoverAction.write-file.task-title")) {
       @Override
       public void run(ProgressIndicator indicator) {
-        Try.run(() -> branchLayoutWriter.write(macheteFilePath, branchLayout, /* backupOldLayout */ true))
-            .onFailure(e -> VcsNotifier.getInstance(project).notifyError(
-                /* title */ getString("action.GitMachete.DiscoverAction.notification.title.write-file-error"),
-                /* message */ e.getMessage() != null ? e.getMessage() : ""))
-            .onSuccess(__ -> baseEnhancedGraphTable.queueRepositoryUpdateAndModelRefresh());
+        Try.run(() -> branchLayoutWriter.write(macheteFilePath, branchLayout, /* backupOldLayout */ true));
+      }
+
+      @Override
+      public void onSuccess() {
+        baseEnhancedGraphTable.queueRepositoryUpdateAndModelRefresh();
+        VfsUtil.markDirtyAndRefresh(/* async */ false, /* recursive */ true, /* reloadChildren */ false,
+            ProjectRootManager.getInstance(project).getContentRoots());
+        postWriteRunnable.run();
```

Comment:
Whoops... does it compile successfully? Coz it's running `@UI Runnable` (a Runnable that can only be executed on UI thread)  on a method that's not explicitly marked as `@UIEffect`... although it could be marked as such

Reply:
> IIRC both these Task's callbacks are executed on UI thread;

I thought that this is why it works

Reply:
Hmmm possibly GuiEffectChecker takes into account `@UIEffect` annotations on base `Task` methods in com.intellij.astub`... still, doesn't harm to mark both `onSuccess` and `onThrowable` as `@UIEffect` for clarity